Implement `INV_TRANSAD` for the GPU branch.

l90lpa opened the pull request:
This draft PR implements `INV_TRANSAD` for the GPU branch. I've reused some existing sudroutines from the direct code, `TRGTOL`, `FTDIR` and `LEDIR`, as on inspection they appear to be the same as the adjoints of `TRLTOG`, `FTINV` and `LEINV`. For `TRMTOL` routines I've added an adjoint as the pack/unpack of `TRLTOM` handle scaling and truncation as well as transposition. Also, I've extended the existing trani `invtrans_adjoint` test to check the vor/div fields, and add a trans test of `INV_TRANSAD` where I've reused some parts of @samhatfield work in https://github.com/ecmwf-ifs/ectrans/pull/214. One thing I haven't done is fix-up the `GSTATS` codes as I'm not exactly sure how the IDs are assigned.

samhatfield commented on src/trans/gpu/external/inv_transad.F90:
```diff
@@ -160,7 +171,463 @@ SUBROUTINE INV_TRANSAD(PSPVOR,PSPDIV,PSPSCALAR,PSPSC3A,PSPSC3B,PSPSC2,&
+
+CALL INV_TRANS_CTLAD(IF_UV_G,IF_SCALARS_G,IF_GP,IF_FS,IF_OUT_LT,&
+ & IF_UV,IF_SCALARS,IF_SCDERS,&
+ & PSPVOR,PSPDIV,PSPSCALAR,KVSETUV,KVSETSC,PGP,FSPGL_PROC,&
```
```suggestion
 & PSPVOR,PSPDIV,PSPSCALAR,KVSETUV,KVSETSC,PGP,&
```

`FSPGL_PROC` isn't used by `INV_TRANSAD` (even though it is an argument) so we can delete this.

samhatfield commented on src/trans/gpu/internal/inv_trans_ctlad_mod.F90:
```diff
@@ -0,0 +1,244 @@
+  CONTAINS
+    SUBROUTINE INV_TRANS_CTLAD(KF_UV_G,KF_SCALARS_G,KF_GP,KF_FS,KF_OUT_LT,&
+      & KF_UV,KF_SCALARS,KF_SCDERS,&
+      & PSPVOR,PSPDIV,PSPSCALAR,KVSETUV,KVSETSC,PGP,FSPGL_PROC,&
```
```suggestion
      & PSPVOR,PSPDIV,PSPSCALAR,KVSETUV,KVSETSC,PGP,&
```

See above.

samhatfield commented on src/trans/gpu/internal/inv_trans_ctlad_mod.F90:
```diff
@@ -0,0 +1,244 @@
+      INTEGER(KIND=JPIM) ,OPTIONAL, INTENT(IN) :: KVSETSC3B(:)
+      INTEGER(KIND=JPIM) ,OPTIONAL, INTENT(IN) :: KVSETSC2(:)
+      REAL(KIND=JPRB) ,OPTIONAL   ,INTENT(IN) :: PGP(:,:,:)
+      EXTERNAL  FSPGL_PROC
+      OPTIONAL  FSPGL_PROC
```
```suggestion
```

See above.

samhatfield commented on src/trans/gpu/external/inv_transad.F90:
```diff
@@ -160,7 +171,463 @@ SUBROUTINE INV_TRANSAD(PSPVOR,PSPDIV,PSPSCALAR,PSPSC3A,PSPSC3B,PSPSC2,&
+ENDIF
+
+IF(PRESENT(KVSETSC)) THEN
+!   IF(.NOT. PRESENT(PSPSCALAR) ) THEN
+!     CALL ABORT_TRANS('INV_TRANSAD : KVSETSC PRESENT BUT PSPSCALAR MISSING')
+!   ENDIF
```
```suggestion
```

Leftover debug code?

samhatfield commented on src/trans/gpu/external/inv_transad.F90:
```diff
@@ -160,7 +171,463 @@ SUBROUTINE INV_TRANSAD(PSPVOR,PSPDIV,PSPSCALAR,PSPSC3A,PSPSC3B,PSPSC2,&
+  NF_SC3B = UBOUND(PSPSC3B,1)
+ENDIF
+
+IF (IF_SCALARS_G > 0 ) THEN
```
In the CPU version we check `IF_SCALARS` so I suggest we do the same here.
```suggestion
IF (IF_SCALARS > 0 ) THEN
```

samhatfield commented on src/trans/gpu/external/inv_transad.F90:
```diff
@@ -160,7 +171,463 @@ SUBROUTINE INV_TRANSAD(PSPVOR,PSPDIV,PSPSCALAR,PSPSC3A,PSPSC3B,PSPSC2,&
+  LUVDER = LDUVDER
+ENDIF
+
+! IF(PRESENT(LDLATLON)) THEN
+!   LATLON = LDLATLON
+! ENDIF
```
```suggestion
```

Leftover debug code?

wdeconinck commented:
Thanks @l90lpa for taking this on!!!

l90lpa commented:
Thanks for taking a first pass @samhatfield, I'll make those fixes next week (and any others as needed). @wdeconinck it's no problem!

samhatfield commented:
> Thanks for taking a first pass @samhatfield, I'll make those fixes next week (and any others as needed). @wdeconinck it's no problem!

I'm just working on a pull request directly to your branch which includes the changes above, so no effort required on your part @l90lpa (other than to merge that PR once it's ready!).

samhatfield commented:
@l90lpa would you mind rebasing this branch against develop? Let me know if there merge conflicts you need help resolving.

l90lpa commented:
> @l90lpa would you mind rebasing this branch against develop? Let me know if there are merge conflicts you need help resolving.

Hi @samhatfield, I've just rebased this branch. I didn't run into any merge conflicts, but I also haven't had time to run the tests myself (I'll be able to do that tomorrow morning).

samhatfield commented:
In my [testing PR](https://github.com/ecmwf-ifs/ectrans/pull/250) build / ci (Debug, linux gnu-13) is failing because the tolerance on the `GPNORM` adjoint test is too low - we can ignore that. The ectrans4py tests are failing like I said earlier. The only concerning failure is for build-hpc / ci-hpc (lumi-g cce) where `ectrans_test_transi_invtrans_adjoint_gpu` is failing. @l90lpa do you know what's going on there? It looks like the RHS and LHS difference is ~1E-08 which seems very large for double precision.

l90lpa commented:
Hi @samhatfield, sorry about the out-of-date develop! I thought I had synced the repo before hand but I guess I hadn't. I'll rebase again and push.

Yes that test failure is concerning. It looks to me like the imaginary part of one of the zero wave number (i.e. `KM=0`) coefficients isn't being set to 0, see:
![](https://github.com/user-attachments/assets/9383f91c-b0db-4fe8-87bb-157026dceb0a)
This make me wonder a couple of things (1) I added a block specifically for setting these to 0 post `LEDIR` (see, https://github.com/JCSDA-internal/ectrans/blob/08f2091bcb3f0c3539d86b282493afbe6c6562e7/src/trans/gpu/internal/ltinvad_mod.F90#L290-L316) perhaps this isn't correct? And, (2) do we actually need that block and instead could be we handle this (and maybe fix this issue) in `PRFI1BAD` by changing the nested loop to,
```
    DO KMLOC=1,D_NUMP
      DO JN=2,R_NSMAX+2
        DO JFLD=1,KFIELDS
          KM = D_MYMS(KMLOC)
          IF (KM /= 0 .AND. JN <= R_NSMAX+2-KM) THEN
            IASM0 = D_NASM0(KM)
            INM = IASM0+((R_NSMAX+2-JN)-KM)*2
            PSPEC(JFLD,INM  ) = PIA(2*JFLD-1,JN+1,KMLOC)
            PSPEC(JFLD,INM+1) = PIA(2*JFLD  ,JN+1,KMLOC)
          ELSEIF (KM == 0) THEN
            IASM0 = D_NASM0(KM)
            INM = IASM0+((R_NSMAX+2-JN)-KM)*2
            PSPEC(JFLD,INM  ) = PIA(2*JFLD-1,JN+1,KMLOC)
            PSPEC(JFLD,INM+1) = 0
          ENDIF
        ENDDO
      ENDDO
    ENDDO
```
Here I have modified the loop bounds to simplify the original if-condition, and then split up the if condition based on whether `KM` is zero.

samhatfield commented:
(Hopefully final) PR opened: https://github.com/JCSDA-internal/ectrans/pull/2.

l90lpa commented:
I see, that makes sense. Thanks for figuring that out, and putting in the fix! I'll merge your PR now.

samhatfield commented:
I think my fix to get the build-hpc / ci-hpc (lumi-g cce) `ectrans_test_transi_invtrans_adjoint_gpu` working by initialising output arrays with zeros may have a slight bug: I am initialising values from `0` to `nvordiv*trans.nspec2` of `rspscalary` to zero, but that array actually has `nscalar*trans.nspec2` elements. `nscalar = 2` and `nvordiv = 1`, hence I'm only zeroing the first half of the array. It's difficult to say from a single invocation of the test if the bug has been fixed because sometimes you get lucky with the RNG. So that's something I need to fix. However at this stage it will be easier to merge this PR and correct the tests afterwards, so that's what I'll do.

Many thanks for the hard work here @l90lpa!
